cmd/snap-seccomp-blacklist/snap-seccomp-blacklist.c

```diff
@@ -0,0 +1,215 @@
+    bool remove_native_arch = true;
+    for (size_t i = 0; i < num_arch_tags; ++i) {
+        if (arch_tags[i] == native_arch) {
+            remove_native_arch = false;
+            break;
+        }
+    }
+    if (remove_native_arch) {
+        sc_err = seccomp_arch_remove(ctx, SCMP_ARCH_NATIVE);
+        if (sc_err < 0) {
+            showerr("cannot remove native architecture");
+            return sc_err;
+        }
+    }
+
+    /* Add 64-bit architectures supported by snapd into the seccomp filter.
+     *
+     * The documentation of seccomp_arch_add() is confusing. It says that after
+     * this call any new rules will be added to this architecture. This is
+     * correct. It doesn't, however, explain that the rules will be multiplied
+     * and re-written as explained below. */
+    for (size_t i = 0; i < num_arch_tags; ++i) {
+        uint32_t arch_tag = arch_tags[i];
+        sc_err = seccomp_arch_add(ctx, arch_tag);
+        if (sc_err < 0 && sc_err != -EEXIST) {
+            showerr("cannot add architecture %x", arch_tag);
+            return sc_err;
+        }
+    }
+
+    /* When the rule set doesn't match one of the architectures above then the
+     * resulting action should be a "allow" rather than "kill". We don't add
+     * any of the 32bit architectures since there is no need for any extra
+     * filtering there. */
+    sc_err = seccomp_attr_set(ctx, SCMP_FLTATR_ACT_BADARCH, SCMP_ACT_ALLOW);
+    if (sc_err < 0) {
+        showerr("cannot set action for unknown architectures");
+        return sc_err;
+    }
+
+    /* Resolve the name of "ioctl" on this architecture. */
+    int sys_ioctl_nr;
+    sys_ioctl_nr = seccomp_syscall_resolve_name("ioctl");
+    if (sys_ioctl_nr == __NR_SCMP_ERROR) {
+        showerr("cannot resolve ioctl system call number");
+        return -ESRCH;
+    }
+
+    /* All of the rules must be added for the native architecture (using native
+     * system call numbers). When the final program is generated the set of
+     * architectures added earlier will be used to determine the correct system
+     * call number for each architecture.
+     *
+     * In other words, As such arguments to scmp_rule_add() must always use
+     * native system call numbers. Translation for the correct architecture
+     * will be performed internally. This is not documented in libseccomp. But
+     * correct operation was confirmed using the pseudo-code program and the
+     * bpf_dbg tool from the kernel tools/bpf directory.
+     *
+     * NOTE: I'm not using scmp_rule_add_exact as that was not doing anything
+     * at all (presumably due to having all the architectures defined). */
+
+    const struct scmp_arg_cmp no_tty_inject = {
+        /* When faced with ioctl(fd, request); refuse to proceed when
+         * request&0xffffffff == TIOCSTI. This specific way to encode the
+         * filter has the following important properties:
+         *
+         * 1) it blocks ioctl(fd, TIOCSTI, ptr).
```

Comment:
I think I'd drop the list, seems unnecessary and quite confusing.

Reply:
I changed the style of the list slightly but I would prefer to retain it unless you can point out what the confusion is about. The list captures the essence of what the program is trying to achieve so in my eyes it is central to what this patch does.